OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs

```diff
@@ -196,10 +210,48 @@ void IBotPositionsUpdated.UpdatedDefenseCenter(CPos newLocation)
 
 		void IBotTick.BotTick(IBot bot)
 		{
+			// TODO: this causes pathfinding lag when AI's gets blocked in
 			SetRallyPointsForNewProductionBuildings(bot);
 
-			foreach (var b in builders)
-				b.Tick(bot);
+			BuildingsBeingProduced = new Dictionary<string, int>();
+
+			// PERF: We tick only one type of valid queue at a time
+			// if AI gets enough cash, AI can make all its queue busy within enough ticks.
+			var findQueue = false;
+			for (int i = 0, builderIndex = currentBuilderIndex; i < builders.Length; i++)
+			{
+				builderIndex = builderIndex < builders.Length - 1 ? builderIndex++ : 0;
+
+				var queues = AIUtils.FindQueues(player, builders[builderIndex].Category).ToArray();
```

Comment:
you can extract `builders[builderIndex]` into a variable, as it's also used in line 251